### PR TITLE
Add telemetry for target type completion

### PR DESCRIPTION
**Pull request by genlu**
#38096
Want to monitor the perf and make sure it doesn't regress when rolling out to broader audience.

@dotnet/roslyn-ide

### CODE REVIEW

File: src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs [hunk 1]
```diff
@@ -462,5 +470,32 @@ protected virtual string GetInsertionText(CompletionItem item, char ch)
         {
             return SymbolCompletionItem.GetInsertionText(item);
         }
+
+        // This is used to decide which provider we'd collect target type completion telemetry from.
+        protected virtual bool ShouldCollectTelemetryForTargetTypeCompletion => false;
```
Comment (mavasani) [inline src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs]: Nit: should this be an abstract property instead?

File: src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs [hunk 2]
```diff
@@ -462,5 +470,32 @@ protected virtual string GetInsertionText(CompletionItem item, char ch)
...
+
+        private class TelemetryCounter : IDisposable
```
Comment (mavasani) [inline src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs]: nit: sealed

File: src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs [hunk 3]
```diff
@@ -462,5 +470,32 @@ protected virtual string GetInsertionText(CompletionItem item, char ch)
...
+            private int _tick;
+
+            public TelemetryCounter(bool shouldReport)
+            {
+                _shouldReport = shouldReport;
+            }
+
+            public void AddTick(int tick)
+            {
+                _tick += tick;
```
Comment (mavasani) [inline src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs]: Just confirming, we are not going to use the counter from from multiple threads, correct?

Comment (genlu) [reply src/Features/Core/Portable/Completion/Providers/AbstractSymbolCompletionProvider.cs]: That's correct.

**Comment by genlu** [comment]
Thanks @mavasani !

**Pull request merged**